File: app/src/main/java/wangdaye/com/geometricweather/remoteviews/presenters/MultiCityWidgetIMP.java
package wangdaye.com.geometricweather.remoteviews.presenters;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.graphics.Color;
import android.util.TypedValue;
import android.view.View;
import android.widget.RemoteViews;

import androidx.annotation.IdRes;
import androidx.annotation.IntRange;

import java.util.List;

import wangdaye.com.geometricweather.GeometricWeather;
import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.background.receiver.widget.WidgetMultiCityProvider;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.options.unit.TemperatureUnit;
import wangdaye.com.geometricweather.common.basic.models.weather.Temperature;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.resource.ResourceHelper;
import wangdaye.com.geometricweather.resource.providers.ResourceProvider;
import wangdaye.com.geometricweather.resource.ResourcesProviderFactory;
import wangdaye.com.geometricweather.settings.SettingsManager;

public class MultiCityWidgetIMP extends AbstractRemoteViewsPresenter {

    public static void updateWidgetView(Context context, List<Location> locationList) {
        WidgetConfig config = getWidgetConfig(
                context,
                context.getString(R.string.sp_widget_multi_city)
        );

        RemoteViews views = getRemoteViews(
                context, locationList,
                config.cardStyle, config.cardAlpha, config.textColor, config.textSize
        );

        AppWidgetManager.getInstance(context).updateAppWidget(
                new ComponentName(context, WidgetMultiCityProvider.class),
                views
        );
    }

    public static RemoteViews getRemoteViews(Context context,
                                             List<Location> locationList,
                                             String cardStyle, int cardAlpha,
                                             String textColor, int textSize) {
        Location location = locationList.get(0);
        Weather weather = location.getWeather();
        boolean dayTime = location.isDaylight();

        ResourceProvider provider = ResourcesProviderFactory.getNewInstance();

        SettingsManager settings = SettingsManager.getInstance(context);
        TemperatureUnit temperatureUnit = settings.getTemperatureUnit();
        boolean minimalIcon = settings.isWidgetMinimalIconEnabled();

        WidgetColor color = new WidgetColor(context, cardStyle, textColor);

        RemoteViews views = new RemoteViews(
                context.getPackageName(),
                !color.showCard
                        ? R.layout.widget_multi_city_horizontal
                        : R.layout.widget_multi_city_horizontal_card
        );

        // city 1.
        views.setViewVisibility(R.id.widget_multi_city_horizontal_weather_1, View.VISIBLE);
        if (weather != null) {
            views.setTextViewText(
                    R.id.widget_multi_city_horizontal_title_1,
                    location.getCityName(context)
            );

            views.setImageViewUri(
                    R.id.widget_multi_city_horizontal_icon_1,
                    ResourceHelper.getWidgetNotificationIconUri(
                            provider,
                            dayTime
                                    ? weather.getDailyForecast().get(0).day().getWeatherCode()
                                    : weather.getDailyForecast().get(0).night().getWeatherCode(),
                            dayTime,
                            minimalIcon,
                            color.getMinimalIconColor()
                    )
            );

            views.setTextViewText(
                    R.id.widget_multi_city_horizontal_content_1,
                    Temperature.getTrendTemperature(
                            context,
                            weather.getDailyForecast().get(0).night().getTemperature().getTemperature(),
                            weather.getDailyForecast().get(0).day().getTemperature().getTemperature(),
                            temperatureUnit
                    )
            );
        }
        setOnClickPendingIntent(context, views, location,
                R.id.widget_multi_city_horizontal_weather_1, 0);

        // city 2.
        if (locationList.size() >= 2) {
            location = locationList.get(1);
            weather = location.getWeather();
            dayTime = location.isDaylight();

            views.setViewVisibility(R.id.widget_multi_city_horizontal_weather_2, View.VISIBLE);
            if (weather != null) {
                views.setTextViewText(
                        R.id.widget_multi_city_horizontal_title_2,
                        location.getCityName(context)
                );

                views.setImageViewUri(
                        R.id.widget_multi_city_horizontal_icon_2,
                        ResourceHelper.getWidgetNotificationIconUri(
                                provider,
                                dayTime
                                        ? weather.getDailyForecast().get(0).day().getWeatherCode()
                                        : weather.getDailyForecast().get(0).night().getWeatherCode(),
                                dayTime,
                                minimalIcon,
                                color.getMinimalIconColor()
                        )
                );

                views.setTextViewText(
                        R.id.widget_multi_city_horizontal_content_2,
                        Temperature.getTrendTemperature(
                                context,
                                weather.getDailyForecast().get(0).night().getTemperature().getTemperature(),
                                weather.getDailyForecast().get(0).day().getTemperature().getTemperature(),
                                temperatureUnit
                        )
                );
            }
            setOnClickPendingIntent(context, views, location,
                    R.id.widget_multi_city_horizontal_weather_2, 1);
        } else {
            views.setViewVisibility(R.id.widget_multi_city_horizontal_weather_2, View.GONE);
        }

        // city 3.
        if (locationList.size() >= 3) {
            location = locationList.get(2);
            weather = location.getWeather();
            dayTime = location.isDaylight();

            views.setViewVisibility(R.id.widget_multi_city_horizontal_weather_3, View.VISIBLE);
            if (weather != null) {
                views.setTextViewText(
                        R.id.widget_multi_city_horizontal_title_3,
                        location.getCityName(context)
                );

                views.setImageViewUri(
                        R.id.widget_multi_city_horizontal_icon_3,
                        ResourceHelper.getWidgetNotificationIconUri(
                                provider,
                                dayTime
                                        ? weather.getDailyForecast().get(0).day().getWeatherCode()
                                        : weather.getDailyForecast().get(0).night().getWeatherCode(),
                                dayTime,
                                minimalIcon,
                                color.getMinimalIconColor()
                        )
                );

                views.setTextViewText(
                        R.id.widget_multi_city_horizontal_content_3,
                        Temperature.getTrendTemperature(
                                context,
                                weather.getDailyForecast().get(0).night().getTemperature().getTemperature(),
                                weather.getDailyForecast().get(0).day().getTemperature().getTemperature(),
                                temperatureUnit
                        )
                );
            }
            setOnClickPendingIntent(context, views, location,
                    R.id.widget_multi_city_horizontal_weather_3, 2);
        } else {
            views.setViewVisibility(R.id.widget_multi_city_horizontal_weather_3, View.GONE);
        }

        if (color.textColor != Color.TRANSPARENT) {
            views.setTextColor(R.id.widget_multi_city_horizontal_title_1, color.textColor);
            views.setTextColor(R.id.widget_multi_city_horizontal_title_2, color.textColor);
            views.setTextColor(R.id.widget_multi_city_horizontal_title_3, color.textColor);
            views.setTextColor(R.id.widget_multi_city_horizontal_content_1, color.textColor);
            views.setTextColor(R.id.widget_multi_city_horizontal_content_2, color.textColor);
            views.setTextColor(R.id.widget_multi_city_horizontal_content_3, color.textColor);
        }

        if (textSize != 100) {
            float titleSize = context.getResources().getDimensionPixelSize(R.dimen.widget_title_text_size)
                    * textSize / 100f;
            float contentSize = context.getResources().getDimensionPixelSize(R.dimen.widget_content_text_size)
                    * textSize / 100f;

            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_title_1, TypedValue.COMPLEX_UNIT_PX, titleSize);
            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_title_2, TypedValue.COMPLEX_UNIT_PX, titleSize);
            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_title_3, TypedValue.COMPLEX_UNIT_PX, titleSize);
            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_content_1, TypedValue.COMPLEX_UNIT_PX, contentSize);
            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_content_2, TypedValue.COMPLEX_UNIT_PX, contentSize);
            views.setTextViewTextSize(R.id.widget_multi_city_horizontal_content_3, TypedValue.COMPLEX_UNIT_PX, contentSize);
        }

        if (color.showCard) {
            views.setImageViewResource(
                    R.id.widget_multi_city_horizontal_card,
                    getCardBackgroundId(color.cardColor)
            );
            views.setInt(
                    R.id.widget_multi_city_horizontal_card,
                    "setImageAlpha",
                    (int) (cardAlpha / 100.0 * 255)
            );
        }

        return views;
    }

    public static boolean isEnable(Context context) {
        int[] widgetIds = AppWidgetManager.getInstance(context)
                .getAppWidgetIds(
                        new ComponentName(context, WidgetMultiCityProvider.class)
                );
        return widgetIds != null && widgetIds.length > 0;
    }

    private static void setOnClickPendingIntent(Context context, RemoteViews views, Location location,
                                                @IdRes int resId, @IntRange(from = 0, to = 2) int index) {
        views.setOnClickPendingIntent(
                resId,
                getWeatherPendingIntent(
                        context,
                        location,
                        GeometricWeather.WIDGET_MULTI_CITY_PENDING_INTENT_CODE_WEATHER_1 + 2 * index
                )
        );
    }
}

File: app/src/main/java/org/breezyweather/remoteviews/presenters/MultiCityWidgetIMP.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.remoteviews.presenters

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.graphics.Color
import android.util.TypedValue
import android.view.View
import android.widget.RemoteViews
import androidx.annotation.IdRes
import androidx.annotation.IntRange
import breezyweather.domain.location.model.Location
import org.breezyweather.R
import org.breezyweather.background.receiver.widget.WidgetMultiCityProvider
import org.breezyweather.domain.location.model.getPlace
import org.breezyweather.domain.location.model.isDaylight
import org.breezyweather.domain.weather.model.getTrendTemperature
import org.breezyweather.remoteviews.Widgets
import org.breezyweather.settings.SettingsManager
import org.breezyweather.theme.resource.ResourceHelper
import org.breezyweather.theme.resource.ResourcesProviderFactory

object MultiCityWidgetIMP : AbstractRemoteViewsPresenter() {

    fun updateWidgetView(
        context: Context,
        locationList: List<Location>,
    ) {
        val config = getWidgetConfig(context, context.getString(R.string.sp_widget_multi_city))
        val views = getRemoteViews(
            context,
            locationList,
            config.cardStyle,
            config.cardAlpha,
            config.textColor,
            config.textSize
        )
        AppWidgetManager.getInstance(context).updateAppWidget(
            ComponentName(context, WidgetMultiCityProvider::class.java),
            views
        )
    }

    fun getRemoteViews(
        context: Context,
        locationList: List<Location>,
        cardStyle: String?,
        cardAlpha: Int,
        textColor: String?,
        textSize: Int,
    ): RemoteViews {
        val provider = ResourcesProviderFactory.newInstance
        val settings = SettingsManager.getInstance(context)
        val temperatureUnit = settings.temperatureUnit
        val minimalIcon = settings.isWidgetUsingMonochromeIcons
        val color = WidgetColor(context, cardStyle!!, textColor!!, locationList.firstOrNull()?.isDaylight ?: true)
        val views = RemoteViews(
            context.packageName,
            if (!color.showCard) R.layout.widget_multi_city_horizontal else R.layout.widget_multi_city_horizontal_card
        )

        val cityIds = arrayOf(
            arrayOf(
                R.id.widget_multi_city_horizontal_weather_1,
                R.id.widget_multi_city_horizontal_title_1,
                R.id.widget_multi_city_horizontal_icon_1,
                R.id.widget_multi_city_horizontal_content_1
            ),
            arrayOf(
                R.id.widget_multi_city_horizontal_weather_2,
                R.id.widget_multi_city_horizontal_title_2,
                R.id.widget_multi_city_horizontal_icon_2,
                R.id.widget_multi_city_horizontal_content_2
            ),
            arrayOf(
                R.id.widget_multi_city_horizontal_weather_3,
                R.id.widget_multi_city_horizontal_title_3,
                R.id.widget_multi_city_horizontal_icon_3,
                R.id.widget_multi_city_horizontal_content_3
            )
        )

        cityIds.forEachIndexed { i, cityId ->
            locationList.getOrNull(i)?.let { location ->
                views.setViewVisibility(cityId[0], View.VISIBLE)
                views.setTextViewText(cityId[1], location.getPlace(context))
                if (location.isDaylight) {
                    location.weather?.dailyForecastStartingToday?.getOrNull(i)?.day?.weatherCode?.let {
                        views.setViewVisibility(cityId[2], View.VISIBLE)
                        views.setImageViewUri(
                            cityId[2],
                            ResourceHelper.getWidgetNotificationIconUri(
                                provider,
                                it,
                                true,
                                minimalIcon,
                                color.minimalIconColor
                            )
                        )
                    } ?: views.setViewVisibility(cityId[2], View.INVISIBLE)
                } else {
                    location.weather?.dailyForecastStartingToday?.getOrNull(i)?.night?.weatherCode?.let {
                        views.setViewVisibility(cityId[2], View.VISIBLE)
                        views.setImageViewUri(
                            cityId[2],
                            ResourceHelper.getWidgetNotificationIconUri(
                                provider,
                                it,
                                false,
                                minimalIcon,
                                color.minimalIconColor
                            )
                        )
                    } ?: views.setViewVisibility(cityId[2], View.INVISIBLE)
                }
                views.setTextViewText(
                    cityId[3],
                    location.weather?.today?.getTrendTemperature(context, temperatureUnit)
                )
                setOnClickPendingIntent(context, views, location, cityId[0], i)
            } ?: views.setViewVisibility(cityId[0], View.GONE)
        }

        if (color.textColor != Color.TRANSPARENT) {
            views.apply {
                setTextColor(R.id.widget_multi_city_horizontal_title_1, color.textColor)
                setTextColor(R.id.widget_multi_city_horizontal_title_2, color.textColor)
                setTextColor(R.id.widget_multi_city_horizontal_title_3, color.textColor)
                setTextColor(R.id.widget_multi_city_horizontal_content_1, color.textColor)
                setTextColor(R.id.widget_multi_city_horizontal_content_2, color.textColor)
                setTextColor(R.id.widget_multi_city_horizontal_content_3, color.textColor)
            }
        }
        if (textSize != 100) {
            val titleSize = context.resources.getDimensionPixelSize(R.dimen.widget_title_text_size)
                .toFloat() * textSize / 100f
            val contentSize = context.resources.getDimensionPixelSize(R.dimen.widget_content_text_size)
                .toFloat() * textSize / 100f
            views.apply {
                setTextViewTextSize(R.id.widget_multi_city_horizontal_title_1, TypedValue.COMPLEX_UNIT_PX, titleSize)
                setTextViewTextSize(R.id.widget_multi_city_horizontal_title_2, TypedValue.COMPLEX_UNIT_PX, titleSize)
                setTextViewTextSize(R.id.widget_multi_city_horizontal_title_3, TypedValue.COMPLEX_UNIT_PX, titleSize)
                setTextViewTextSize(
                    R.id.widget_multi_city_horizontal_content_1,
                    TypedValue.COMPLEX_UNIT_PX,
                    contentSize
                )
                setTextViewTextSize(
                    R.id.widget_multi_city_horizontal_content_2,
                    TypedValue.COMPLEX_UNIT_PX,
                    contentSize
                )
                setTextViewTextSize(
                    R.id.widget_multi_city_horizontal_content_3,
                    TypedValue.COMPLEX_UNIT_PX,
                    contentSize
                )
            }
        }
        if (color.showCard) {
            views.setImageViewResource(R.id.widget_multi_city_horizontal_card, getCardBackgroundId(color))
            views.setInt(R.id.widget_multi_city_horizontal_card, "setImageAlpha", (cardAlpha / 100.0 * 255).toInt())
        }
        return views
    }

    fun isInUse(context: Context): Boolean {
        val widgetIds = AppWidgetManager.getInstance(context)
            .getAppWidgetIds(ComponentName(context, WidgetMultiCityProvider::class.java))
        return widgetIds != null && widgetIds.isNotEmpty()
    }

    private fun setOnClickPendingIntent(
        context: Context,
        views: RemoteViews,
        location: Location,
        @IdRes resId: Int,
        @IntRange(from = 0, to = 2) index: Int,
    ) {
        views.setOnClickPendingIntent(
            resId,
            getWeatherPendingIntent(
                context,
                location,
                Widgets.MULTI_CITY_PENDING_INTENT_CODE_WEATHER_1 + 2 * index
            )
        )
    }
}
